[ylchin] iP

ylchin opened the pull request:
# Elara

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

Elara frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- FAST SUPER FAST to use

All you need to do is,

1. download it from [here](https://github.com/nus-cs2103-AY2425S1/ip/pull/469).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉
And it is **FREE**!

Features:

- [X] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the main method:

```java
 public static void main(String[] args) {
        new Elara().run();
    }
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

ylchin committed:
Renamed main class

ylchin committed:
Added greeting

ylchin committed:
Added support for echo

ylchin committed:
Added ability to add and list

ylchin committed:
Added new Task class

ylchin committed:
fixed bugs in Task class

ylchin committed:
Add the ability to mark tasks as done and change it back to undone

ylchin committed:
Updated Task class, added DeadlineTask, EventTask, ToDoTask classes

ylchin committed:
Add support for tracking three types of tasks: ToDos, Deadlines and Events

ylchin committed:
Automated text UI testing

ylchin committed:
Created classes for exceptions regarding the creation of the different task types

ylchin committed:
Added error handling

ylchin committed:
Added support for catching when index is out of range

ylchin committed:
Added support for deleting tasks from the list.

ylchin committed:
Add method to format tasks to write to file

The different task classes did not support returning a String suitable for writing to a file to support data storage.

Fixing this will allow us to write to a file to support storage.

We add the required constructors that enable setting of boolean isDone and method toFileFormat()

ylchin committed:
Support reading from and writing to file

Elara did not support reading to and writing from file, so user has to
retype todo list every time Elara is run

Restructuring Elara by adding methods to read files, parse files for
tasks and save tasks to file allows us to store the data in a specified
file.

Let's reorganise code for Elara by breaking up into smaller blocks
and add methods readFile(), setupFile(), parseTask() and
saveTasksToFile().

ylchin committed:
Update DeadlineTaskException and EventException error message

Previous error message did not specify format for deadline and start and end
time for deadline and event tasks.

Updating the error messages allows users to know the proper format to send into
Elara such that they can efficiently use the task list.

Let's update the error message to include the format to parse for LocalDateTime

ylchin committed:
Change format for deadlines and event start and end to LocalDateTime

The previous type for storing deadlines and event timings was String, which
does not allow for more meaningful storage.

Changing the type to LocalDateTime allows the chatbot to understand that
it is a date and time instead of just a String.

Let's edit the DeadlineTask and EventTask variables to reflect this, as well
as change the format of parseTask() to support reading such data from files.

ylchin committed:
Refactor code to extract out closely related code as classes

Main class Elara is very long and messy due to a lot of similar code.

Splitting the code into different classes allows for modularisation and easier debugging.

Let's implement the classes and Command interface.

ylchin committed:
Divide classes into packages

Classes are currently located in the root package without clear
organization.

This unstructured organization makes the codebase harder to navigate
and maintain. It violates OOP principles of encapsulation and
modularity, making it difficult to locate related classes and
understand the overall architecture.

Divide classes into specific packages based on their functionality:
- Move command-related classes (`Command`, `AddCommand`, `ListCommand`,
  etc.) into a `command` package.
- Place task-related classes (`Task`, `ToDoTask`, `DeadlineTask`,
  `EventTask`) and exception classes (`InvalidInputException`,
  `ToDoException`, `DeadlineException`, `EventException`) into a `task`
  package.
- Organize file handling classes (`Storage`) into a `storage` package.
- Group user interaction classes (`Ui`) into a `ui` package.
- Consolidate parsing logic (`Parser`) into a `parser` package.

This organization follows OOP principles, promoting encapsulation and
modularity. It makes the codebase more readable, maintainable, and
scalable by logically grouping related functionalities. Future
development and troubleshooting will be easier, as classes will be
clearly categorized and located.

No functional changes are introduced; this refactor strictly improves
the project structure.

ylchin committed:
Merge commit '81a9c537afe8492675eb253968939364d65db091'

ylchin committed:
Automate project builds using Gradle

Project builds are currently managed manually without a standardized
build automation tool.

This manual process is error-prone and inefficient, making it harder
to manage dependencies, compile code, run tests, and package
distributions consistently. It slows down development and increases
the likelihood of build-related errors.

Automate project builds using Gradle:
- Add a `build.gradle` file to define project dependencies and build
  configurations.
- Include Gradle wrapper files to standardize the Gradle version used
  across different development environments.
- Configure tasks for compiling code, running tests, and packaging the
  application into distributable formats.

Using Gradle provides a standardized build process, improving
consistency and reliability. It simplifies dependency management,
reduces setup time for new developers, and facilitates integration
with CI/CD pipelines.

This change ensures that all developers and CI systems use the same
build process, enhancing collaboration and project stability.

ylchin committed:
Revert "Automate project builds using Gradle"

This reverts commit 4ed9ecb90add2465779d65f31045ff56d19871ec.

ylchin committed:
Update build.gradle file

ylchin committed:
Fixed missing import and package statements

ylchin committed:
Merge commit '55b7f723c5dd4de5427e753d44ac62cc2f471beb'

ylchin committed:
Fix bug to catch if no argument

ylchin committed:
Fixed bug in AddCommand class

ylchin committed:
Add JUnit tests

ylchin committed:
Package the App as a JAR file

Rachelcoll commented on src/main/java/elara/command/AddCommand.java:
```diff
@@ -0,0 +1,98 @@
+    public void execute(TaskList taskList, Ui ui, Storage storage) throws InvalidInputException {
+        Task newTask = null;
+
+        switch (commandType) {
+            case "todo":
+                if (details == null || details.isEmpty()) {
+                    throw new ToDoException();
+                }
+                newTask = new ToDoTask(details);
+                break;
+            case "deadline":
```
Is the switch and case indentation here supposed to be the same according to the coding standard?

ylchin replied:
fixed in later commits

Rachelcoll commented on src/main/java/elara/command/DeleteCommand.java:
```diff
@@ -0,0 +1,30 @@
+    }
+
+    @Override
+    public void execute(TaskList taskList, Ui ui, Storage storage) throws InvalidInputException {
```
Would it be better to add a JavaDoc above this method?

ylchin replied:
have added in later commits

Rachelcoll commented on src/main/java/elara/task/DeadlineTask.java:
```diff
@@ -0,0 +1,24 @@
+    @Override
+    public String toString() {
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("MMM d yyyy hh:mm a");
+        return String.format("[D]%s (by: %s)", super.toString(), deadline.format(formatter));
```
I like the usage of string formatting here! 👍

Rachelcoll commented on src/main/java/elara/storage/Storage.java:
```diff
@@ -0,0 +1,84 @@
+package elara.storage;
+
+import elara.task.DeadlineTask;
+import elara.task.EventTask;
+import elara.task.Task;
+import elara.task.TaskList;
+import elara.task.ToDoTask;
+
+import java.io.File;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+import java.util.ArrayList;
+import java.util.Scanner;
```
Perhaps java packages should be imported before self-defined packages?

ylchin replied:
fixed!

Rachelcoll reviewed (commented):
LGTM, only some nits on coding standard violations to be fixed.

ylchin committed:
Fixed bug DeleteCommand MarkCommand and UnmarkCommand classes

ylchin committed:
Add JavaDoc comments

The current codebase lacks proper JavaDoc comments, which
makes it difficult for developers to understand the purpose
and behavior of classes and methods.

This needs to change because clear documentation improves
code readability and maintainability, and helps onboard
new contributors more easily.

These comments describe the purpose of classes, methods,
and parameters in detail, aligning with standard JavaDoc
conventions.

This approach was chosen to enhance code comprehension
and ensure future maintainability. It provides clear
guidelines for developers using or modifying the code.

No functional changes have been made to the application;
this commit focuses purely on documentation.

ylchin committed:
Fixed bug in MarkCommand DeleteCommand UnmarkCommand classes

ylchin committed:
Fixed bug in MarkCommand DeleteCommand UnmarkCommand classes

ylchin committed:
Edit the code to comply with Java coding standards

The current code does not fully comply with Java coding standards,
particularly in variable naming conventions such as the usage of
camelCase for private final variables.

This needs to change to improve code consistency, readability, and
ensure it follows widely accepted Java conventions, making it easier
for developers to understand and maintain.

Refactored variable names like `FILE_PATH` to use camelCase as per
Java standards. Private final variables now adhere to the camelCase
convention.

This approach ensures the codebase is easier to read and maintain,
and that it conforms to industry-standard Java naming practices.

No functional changes were made; this commit is purely a style
refactor for standardization.

ylchin committed:
Add find command for searching tasks by keyword

Currently, there is no way for users to search for tasks
based on a keyword in the task description. This makes it
difficult to find specific tasks quickly, especially when
there are many tasks.

This needs to change to improve the user experience by
allowing users to filter tasks based on a keyword, making
task management more efficient.

A new FindCommand class has been implemented to handle the
'find' command. The Parser has been updated to recognize the
'find' command and parse the search keyword. A findTasks
method has been added to the TaskList class to return tasks
matching the keyword. The Ui class now includes a method
to display the matching tasks to the user.

This approach was chosen to align with the existing command
pattern, keeping the code modular and easily extensible.

ylchin committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/elara/command/AddCommand.java
#	src/main/java/elara/command/DeleteCommand.java
#	src/main/java/elara/command/MarkCommand.java
#	src/main/java/elara/storage/Storage.java
#	src/main/java/elara/ui/Ui.java

ylchin committed:
Resolve merge conflict in AddCommand, DeleteCommand,
MarkCommand, UnmarkCommand, Storage and Ui classes

A merge conflict occurred during the integration of changes
from different branches. This resulted in conflicting modifications
in [file names], which prevented the merge from completing.

This needs to change because the merge conflict must be resolved
to integrate the changes from both branches and maintain consistency
in the codebase.

The conflicting sections in the files were manually reviewed and
merged. The resolved code incorporates the necessary changes from
both branches, ensuring no loss of functionality.

This approach ensures the smooth integration of changes and restores
the ability to continue development on a clean codebase.

No additional changes were made beyond resolving the conflict.

ylchin committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/elara/command/AddCommand.java
#	src/main/java/elara/command/MarkCommand.java
#	src/main/java/elara/command/UnmarkCommand.java
#	src/main/java/elara/parser/Parser.java
#	src/main/java/elara/ui/Ui.java

ylchin committed:
Edit import order and line breaks to comply with Java coding standards following merge

The recent merge introduced inconsistent import ordering
and line breaks that do not adhere to Java coding standards.

This needs to change to maintain consistent code formatting
and readability throughout the project, ensuring compliance
with Java best practices.

Reordered import statements in AddCommand, DeleteCommand,
MarkCommand, UnmarkCommand, and Ui classes to follow
convention. Additionally, adjusted line breaks to improve
code readability and ensure consistency across files.

This approach ensures uniformity in code structure and
enhances readability without altering functionality.

ylchin committed:
Add Javadocs that were lost following merge in TaskList.java

During the recent merge, Javadoc comments in
TaskList.java were unintentionally lost. This
resulted in incomplete documentation of the
class and its methods.

This needs to be fixed to ensure proper
documentation is restored, aiding developers
in understanding the purpose and functionality
of the TaskList class and its methods.

Re-added Javadocs for all constructors and
methods in TaskList.java, providing clear
explanations of their usage and parameters.

This approach ensures the code remains well
documented and consistent, helping future
maintenance and improving readability for
other developers.

ylchin committed:
Configure Checkstyle for code quality checks

Currently, there is no automated way to enforce coding standards in the project.

To ensure consistency and detect coding style violations, Checkstyle is introduced.
The configuration files (checkstyle.xml and suppressions.xml) are added under config/checkstyle,
and necessary changes are made to the build.gradle file to integrate Checkstyle into the build process.

This will allow the project to maintain a high standard of code quality and detect issues early during development.

ylchin committed:
Refactor code to comply with Checkstyle standards

Currently, the code does not fully adhere to the coding standards enforced by Checkstyle,
which has led to violations during code quality checks.

This commit refactors the code to ensure it complies with the Checkstyle configuration.
Changes include adjustments to import order, line breaks, spacing, as well as the addition of necessary Javadoc comments.

Ensuring compliance with Checkstyle helps maintain a consistent coding style across the project and improves readability and maintainability.

ylchin committed:
Update build.gradle to support JavaFX integration

The current build.gradle configuration does not include support for JavaFX,
which is required for the graphical user interface components of the project.

This commit updates the build.gradle file to include the necessary dependencies
and configurations for JavaFX. It adds the JavaFX plugin and specifies the version
and modules required for the project.

This ensures that the project can successfully compile and run JavaFX applications,
enabling the development of rich desktop applications with a graphical interface.

ylchin committed:
Add FXML files for JavaFX interface components

Currently, the project lacks the necessary FXML files to define the structure of the JavaFX interface.

This commit adds `DialogBox.fxml` and `MainWindow.fxml` to the resources/view directory,
which define the layout for the dialog boxes and the main window interface in the JavaFX application.

This ensures that the project has the required layout files to support JavaFX-based user interface components.

ylchin committed:
Add images for user and chatbot icons in the GUI

The current GUI lacks visual representation for the user and the chatbot.

This commit adds `Elara.jpg` and `User.jpg` to the resources/images directory. These images
will be used as icons representing the chatbot (Elara) and the user in the JavaFX GUI dialogs.

This enhances the user interface by adding visual cues, improving user experience and interaction clarity.

ylchin committed:
Refactor package structure and update Ui methods

Files like `Parser.java` and `Storage.java` are scattered across
different packages, leading to inconsistencies in project structure.

To maintain a clean and organized codebase, files related to command
execution and components should be grouped in one package.

The files `Parser.java`, `Storage.java`, and `UnmarkCommand.java` have
been moved to the `components` package. Additionally, the methods in
`Ui.java` that previously printed messages directly have been refactored
to return strings instead. This enhances flexibility and supports better
unit testing by decoupling UI logic from direct console printing.

This approach improves maintainability and allows for better
modularization of the system.

ylchin committed:
Revert "Add images for user and chatbot icons in the GUI"

This reverts commit 53a7a7749ef7adb75ddce7cfd9ade05aec94240d.

ylchin committed:
Refactor package structure and update Ui methods

The current situation involves files spread across different packages,
leading to challenges in maintaining consistency across imports and
method calls.

To streamline and centralize the code, files from various packages were
moved to a unified utils package. This simplifies imports and enhances
maintainability by keeping related components together.

Additionally, methods in Ui.java were updated to return strings
instead of printing directly. This approach improves flexibility in
handling output, making it easier to test and reuse components.
The task.getDescription() method was also added to support these
changes.

These changes were made to ensure consistency in the code structure
and improve the system’s modularity.

ylchin committed:
Update methods in Ui.java

The current situation is that the `Ui` class returns strings for
displayed messages rather than printing them directly, which limits
flexibility in how these messages are used or tested.

The code needs to be updated to improve flexibility by displaying
messages directly instead of returning them as strings. This change
allows for more seamless interaction within the program and centralizes
the logic related to handling outputs.

To address this, I modified methods in the `Ui` class to print messages
directly to the console rather than returning them as strings. I also
introduced a `lastResponse` field to keep track of the most recent
message for testing or further use.

This approach enhances clarity and makes the user interface code more
streamlined while allowing access to the last printed message for
testing purposes or other internal uses.

ylchin committed:
GravatarAdd images for user and chatbot icons in the GUI

The current GUI lacks visual representation for the user and the chatbot.

This commit adds `Elara.jpg` and `User.jpg` to the resources/images directory. These images
will be used as icons representing the chatbot (Elara) and the user in the JavaFX GUI dialogs.

This enhances the user interface by adding visual cues, improving user experience and interaction clarity.

ylchin committed:
Add newline at the end of TaskList.java

The TaskList.java file was missing a newline at the end, which is a
standard coding convention for better readability and to avoid issues
with version control systems that may flag it as a change.

A newline has been added at the end of TaskList.java to comply with
Java coding standards and maintain consistency across files.

This change ensures that the file conforms to best practices.

ylchin committed:
Add new JavaFX components for GUI support

The following files were added to support the graphical user interface
for Elara:

- DialogBox.java: Defines the structure and behavior for chat dialog
  boxes in the UI.
- Launcher.java: Handles the application launch.
- Main.java: Entry point for the JavaFX application.
- MainWindow.java: Contains the layout and control logic for the
  main window.

These additions provide the foundational components needed for the
Elara chatbot's graphical interface, enhancing user interaction
and experience.

ylchin committed:
Use assert feature to document important assumptions in the code

Assertions have been added throughout the codebase to document and enforce
critical assumptions that should hold true during execution. This provides
a safeguard to catch unexpected issues by verifying key assumptions such
as non-null variables, valid input formats, and task data consistency.

The affected files include:
- build.gradle (to ensure assertions are enabled during development)
- MainWindow.java
- DeadlineTask.java
- EventTask.java
- ToDoTask.java
- Parser.java

Assertions are a valuable tool for catching logical errors and ensuring
the integrity of the application during runtime.

ylchin committed:
Refactor code to improve overall quality

Critically examined and refactored various parts of the codebase to
enhance code quality.

- Refactored methods to improve readability and maintainability.
- Improved exception handling to ensure robustness.
- Enhanced Javadoc comments for better documentation.
- Simplified logic in certain areas to reduce complexity.
- Organized import statements and optimized code structure.

These changes aim to make the code more readable, maintainable, and
efficient, thus improving the overall quality of the codebase.

ylchin committed:
Merge pull request #2 from ylchin/branch-A-Assertions

Use assert feature to document important assumptions in the code

ylchin committed:
Merge pull request #3 from ylchin/branch-A-CodeQuality

Refactor code to improve overall quality

ylchin committed:
Handle duplicate detection for ToDoTask, DeadlineTask, and EventTask

Currently, there is no mechanism to detect and handle duplicate
tasks across different task types like ToDoTask, DeadlineTask,
and EventTask. This can lead to multiple identical tasks being
added to the list, which can cause confusion and clutter.

This change implements duplicate detection by overriding the
`equals` and `hashCode` methods for ToDoTask, DeadlineTask,
and EventTask.

- For `ToDoTask`, duplicates are identified based solely on the
  task description.
- For `DeadlineTask`, duplicates are identified by matching both
  the description and the deadline.
- For `EventTask`, duplicates are detected by matching the
  description, start time, and end time.

This ensures that tasks are uniquely identified not just by their
description but also by their associated date and time, which is
essential for managing deadlines and events accurately.

This approach maintains code clarity and ensures consistency in
task handling across different task types.

ylchin committed:
Fix bug in duplicate detection where duplicate tasks were still added

Previously, the duplicate detection mechanism failed to prevent
the addition of tasks that were already present in the task list.
This allowed duplicate tasks to be added despite detection
logic, leading to clutter and redundancy in the task list.

This fix ensures that when a duplicate task is detected, it is
not added to the task list. Updated the logic to verify if a task
already exists before adding it to the list for ToDoTask,
DeadlineTask, and EventTask.

Now, if a user attempts to add a duplicate task, the task will
not be added, and a message will be displayed to inform the
user of the duplicate entry.

This change ensures better management of tasks by maintaining
a clean, duplicate-free list.

ylchin committed:
feat(docs): Add product screenshot and update README

The documentation does not include a visual representation of
the product or the updated README content.

Including a product screenshot in the documentation enhances
user understanding of the product and provides a quick visual
reference. The README file likely needs updates to reflect the
current state of the project.

- Added a product screenshot (`product screenshot.png`) to the
  `docs` directory.
- Updated `README.md` to include relevant information, possibly
  including the new screenshot.

The screenshot was added to the `docs` folder for easy
reference and to maintain a centralized location for
documentation assets. The README was updated to improve user
guidance and project overview.

ylchin committed:
refactor(docs): Rename product screenshot.png to Ui.png

The screenshot image file is named product screenshot.png,
which is not descriptive and uses spaces in the filename.

Renaming the file to Ui.png provides a clearer indication
that the image represents the UI. Removing spaces in the
filename improves file handling across different systems.

- Renamed product screenshot.png to Ui.png in the docs
directory.

The new name is concise and better reflects the content of the
image, making it easier to identify and reference in the
documentation.

ylchin committed:
fix(gui): Show welcome message on application startup

The welcome message does not appear in the GUI when the application
is started.

Users should see the welcome message upon launching the application
to enhance user experience and provide initial guidance.

- Modified Elara.java and MainWindow.java to ensure the welcome
message is displayed in the GUI when the app starts.

The changes ensure the welcome message is correctly passed to and
displayed in the main window upon initialization, providing immediate
feedback to users.

ylchin committed:
feat(gui): Update GUI to auto-resize elements and update Ui.png

The GUI elements do not automatically resize, which can cause layout
issues on different screen sizes. The Ui.png file is outdated and
does not reflect the latest design.

Auto-resizing ensures the application looks consistent and
user-friendly across various screen sizes. Updating the Ui.png file
provides an accurate visual reference.

- Modified MainWindow.fxml to enable auto-resizing for GUI elements.
- Updated docs/Ui.png to reflect the current design of the GUI.
- Made minor adjustments in README.md to align with the new
screenshots.

Ensuring auto-resize provides a more flexible UI layout that adjusts
to different screen dimensions, enhancing usability.

ylchin committed:
Update User Guide
